### PR TITLE
validation: return more helpful results for reconsiderable fee failures and skipped transactions

### PR DESCRIPTION
**Pull request by glozow**
Split off from #26711 (suggested in https://github.com/bitcoin/bitcoin/pull/26711#issuecomment-1786392253). This is part of #27463.

- Add 2 new TxValidationResults
  - `TX_RECONSIDERABLE` helps us encode transactions who have failed fee checks that can be bypassed using package validation. This is distinguished from `TX_MEMPOOL_POLICY` so that we re-validate a transaction if and only if it is eligible for package CPFP. In the future, we will have a separate cache for reconsiderable rejects so these transactions don't go in `m_recent_rejects`.
  - `TX_UNKNOWN` helps us communicate that we aborted package validation and didn't finish looking at this transaction: it's not valid but it's also not invalid (i.e. don't cache it as a rejected tx)
- Return effective feerate and the wtxids of transactions used to calculate that effective feerate when the error is `TX_SINGLE_FAILURE`. Previously, we would only provide this information if the transaction passed. Now that we have package validation, it's much more helpful to the caller to know how the failing feerate was calculated. This can also be used to improve our submitpackage RPC result (which is currently a bit unhelpful when things fail).
- Use the newly added `CheckPackageMempoolAcceptResult` for existing package validation tests. This increases test coverage and helps test the changes made in this PR.

### CODE REVIEW

**Comment by instagibbs** [comment]
> so that we re-validate a transaction if and only if it is eligible for package CPFP.

and in the future we'll use this for making sure we'd re-request this transaction as part of another package at p2p, correct?

**Comment by glozow** [comment]
> and in the future we'll use this for making sure we'd re-request this transaction as part of another package at p2p, correct?

Correct :+1:

**Comment by instagibbs** [comment]
think you'll have to touch `CheckATMPInvariants` in the tx_pool fuzz test to add `TX_SINGLE_FAILURE` case

File: src/validation.cpp [hunk 1]
```diff
@@ -1139,7 +1147,7 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
     if (!args.m_package_submission && !bypass_limits) {
         LimitMempoolSize(m_pool, m_active_chainstate.CoinsTip());
         if (!m_pool.exists(GenTxid::Txid(hash)))
-            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "mempool full");
+            return state.Invalid(TxValidationResult::TX_SINGLE_FAILURE, "mempool full");
```
Comment (instagibbs) [inline src/validation.cpp:1150]: Might make sense to add a comment to the effect that we don't want to reject fetching this same tx in a future package

File: src/validation.cpp [hunk 2]
```diff
@@ -1227,7 +1235,13 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     Workspace ws(ptx);
 
-    if (!PreChecks(args, ws)) return MempoolAcceptResult::Failure(ws.m_state);
+    if (!PreChecks(args, ws)) {
+        if (ws.m_state.GetResult() == TxValidationResult::TX_SINGLE_FAILURE) {
+            // Failed for fee reasons. Provide the effective feerate and which txns were included.
```
Comment (instagibbs) [inline src/validation.cpp:1240]:
```suggestion
            // Failed for fee reasons. Provide the effective feerate and which txn was included.
```

File: src/validation.h [hunk 3]
```diff
@@ -141,24 +160,28 @@ struct MempoolAcceptResult {
...
     /** The wtxid of the transaction in the mempool which has the same txid but different witness. */
     const std::optional<uint256> m_other_wtxid;
 
     static MempoolAcceptResult Failure(TxValidationState state) {
         return MempoolAcceptResult(state);
     }
 
+    static MempoolAcceptResult FeeFailure(TxValidationState state,
+                                          CFeeRate effective_feerate,
+                                          const std::vector<uint256>& wtxids_fee_calculations) {
```
Comment (instagibbs) [inline src/validation.h:181]: `Wtxid`? :point_right: :point_left:

**Review by instagibbs — COMMENTED** [review]
LGTM with a few suggestions https://github.com/bitcoin/bitcoin/pull/28785/commits/6b0dc669886c07a81fdfbf2d1a71a77ca3bac1c3

File: src/validation.cpp (hunk 2 above)
Comment (glozow) [reply src/validation.cpp]: Taken

File: src/validation.h (hunk 3 above)
Comment (glozow) [reply src/validation.h]: Changed :+1:

File: src/consensus/validation.h [hunk 4]
```diff
@@ -54,6 +54,8 @@ enum class TxValidationResult {
     TX_CONFLICT,
     TX_MEMPOOL_POLICY,        //!< violated mempool's fee/size/descendant/RBF/etc limits
     TX_NO_MEMPOOL,            //!< this node does not have a mempool so can't validate the transaction
+    TX_SINGLE_FAILURE,        //!< fee was insufficient to meet some policy, but can change if submitted in a (different) package
```
Comment (murchandamus) [inline src/consensus/validation.h:57]: It was not obvious to me what the motivation for `SINGLE` in the constant name here is. After discussing with @glozow, maybe:

```suggestion
    TX_RECONSIDERABLE,        //!< mining score was insufficient to meet some policy, but transaction might be acceptable if submitted in a (different) package
```

Comment (glozow) [reply src/consensus/validation.h]: Yeah renamed, I think it'll help make the connection with `m_reconsiderable_rejects` filter later as well.

File: src/test/fuzz/tx_pool.cpp [hunk 5]
```diff
@@ -154,12 +154,15 @@ void CheckATMPInvariants(const MempoolAcceptResult& res, bool txid_in_mempool, b
         // It may be already in the mempool since in ATMP cases we don't set MEMPOOL_ENTRY or DIFFERENT_WITNESS
         Assert(!res.m_state.IsValid());
         Assert(res.m_state.IsInvalid());
+
+        const bool is_single_failure{res.m_state.GetResult() == TxValidationResult::TX_RECONSIDERABLE};
```
Comment (murchandamus) [inline src/test/fuzz/tx_pool.cpp:158]: After the name change, perhaps:

```suggestion
        const bool is_reconsiderable{res.m_state.GetResult() == TxValidationResult::TX_RECONSIDERABLE};
```

File: src/test/util/txmempool.cpp [hunk 6]
```diff
@@ -89,11 +89,14 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
         }
 
         // m_effective_feerate and m_wtxids_fee_calculations should exist iff the result was valid
-        if (atmp_result.m_effective_feerate.has_value() != valid) {
+        // or if the failure was TX_RECONSIDERABLE
+        const bool valid_or_single_failure{atmp_result.m_result_type == MempoolAcceptResult::ResultType::VALID ||
```
Comment (murchandamus) [inline src/test/util/txmempool.cpp:93]: Perhaps:
```suggestion
        const bool valid_or_reconsiderable{atmp_result.m_result_type == MempoolAcceptResult::ResultType::VALID ||
```

**Review by murchandamus — COMMENTED** [review]
crACK 0150e860863f95b18448e7b67f5db27017660670 with grain of salt:

I’m pretty new to reviewing mempool code:

- The introduction of an explicit class for `Wtxid` makes sense to me
- The introduction of `TX_RECONSIDERABLE` makes sense to me in the context of transactions being tested in the context of packages where they might be resubmitted with a higher resulting mining score
- It makes sense to me that some mempool acceptance tests would now fail with `TX_RECONSIDERABLE` and the ones that do look reasonable to me
- The code changes look reasonable small and targeted to the context

I cannot tell whether the changes are complete and whether they are being comprehensively tested.

File: src/test/fuzz/tx_pool.cpp (hunk 5 above)
Comment (ismaelsadeeq) [reply src/test/fuzz/tx_pool.cpp:158]: PR description also needs to be updated to `TX_RECONSIDERABLE`

Comment (glozow) [reply src/test/fuzz/tx_pool.cpp]: taken

File: src/test/util/txmempool.cpp (hunk 6 above)
Comment (glozow) [reply src/test/util/txmempool.cpp]: taken

**Comment by glozow** [comment]
Addressed https://github.com/bitcoin/bitcoin/pull/28785#pullrequestreview-1716255880 and fixed the ci error

**Comment by instagibbs** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/28785/commits/1147e00e59e47f27024ec96629993c66a3ce4ef0

took reconsiderable variable renames, tests using CheckPackageMempoolAcceptResult more, unsure why ci was failing though :)

**Comment by achow101** [comment]
ACK 1147e00e59e47f27024ec96629993c66a3ce4ef0

**Review by murchandamus — COMMENTED** [review]
reACK 1147e00e59e47f27024ec96629993c66a3ce4ef0

**Review by ismaelsadeeq — APPROVED** [review]
ACK 1147e00e59e47f27024ec96629993c66a3ce4ef0

looks good to me.

**Pull request merged**